pywebdriver/plugins/mettler_toledo_8217.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-FileCopyrightText: 2022 Coop IT Easy SCRLfs
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import logging
+import re
+
+import serial
+
+from .scale_driver import AbstractScaleDriver
+
+ANSWER_RE = re.compile(rb"^\?(?P<status>.)|(?P<weight>\d+\.\d+)$")
+
+_logger = logging.getLogger(__name__)
+
+
+class MettlerToledo8217ScaleDriver(AbstractScaleDriver):
+    """Driver for the 8217 Mettler Toledo protocol. Because of Python
+    restrictions, the number doesn't come first in the class name.
+    """
+
+    def __init__(self, config, *args, **kwargs):
+        super().__init__(config, *args, **kwargs)
+        self.vendor_product = "mettler_toledo_8217"
+
+    @property
+    def _port(self):
+        return self.config.get("scale_driver", "port", fallback="/dev/ttyS0")
+
+    @property
+    def _baudrate(self):
+        return self.config.getint("scale_driver", "baudrate", fallback=9600)
+
+    def acquire_data(self, connection):
+        """Acquire data over the connection."""
+        buffer = b""
+        stx = False
+        # ask for weight data
+        connection.write(b"W")
+        while True:
+            c = connection.read(1)
+            if not c:
+                # timeout
+                raise serial.SerialTimeoutException()
+            if c == b"\x02":
+                # start of answer
+                stx = True
+                buffer = b""
+            elif c == b"\r":
+                # end of answer
+                if not stx:
+                    continue
+                break
+            else:
+                buffer += c
+        match = ANSWER_RE.match(buffer)
+        if match is None:
+            raise ValueError("serial readout was not valid")
+        matchdict = match.groupdict()
+        _logger.debug(matchdict)
+        status = matchdict["status"]
+        weight = matchdict["weight"]
+        result = self.data.copy()
```

Comment:
with the latest changes, sub-classes should not access `self.data`. a new dict should be returned each time.

Reply:
A new dict is returned each time. The old dict is never edited.

The reason we copy the previous dict is because we don't always get a value for `weight` from acquire_data, meaning that we have to fallback on _something_. I chose to fallback on the previous `weight` value instead of `0`. I don't know which is preferable.

Reply:
in [odoo’s iot box code](https://github.com/coopiteasy/OCB/blob/12.0-posbox-version-before-iot-refactoring/addons/hw_scale/controllers/main.py#L28), it seems to return `None` for weight (or `0.0` for a negative weight) when the scale returns a non-weight response. i think it’s cleaner to not access `self.data`. in [odoo’s pos js code](https://github.com/coopiteasy/OCB/blob/12.0/addons/point_of_sale/static/src/js/screens.js#L396), if weight in null, it is set to 0. if in doubt, you can launch posboxless with odoo’s scale driver and inspect the returned values.